addons/binding/org.openhab.binding.nest/src/main/java/org/openhab/binding/nest/internal/rest/NestStreamingRestClient.java

```diff
@@ -0,0 +1,167 @@
+/**
+ * Copyright (c) 2010-2017 by the respective copyright holders.
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.openhab.binding.nest.internal.rest;
+
+import java.time.Duration;
+import java.util.List;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+
+import javax.net.ssl.SSLContext;
+import javax.ws.rs.client.Client;
+import javax.ws.rs.client.ClientBuilder;
+
+import com.google.gson.GsonBuilder;
+import org.glassfish.jersey.SslConfigurator;
+import org.glassfish.jersey.media.sse.EventSource;
+import org.glassfish.jersey.media.sse.InboundEvent;
+import org.glassfish.jersey.media.sse.SseFeature;
+import org.openhab.binding.nest.internal.data.TopLevelData;
+import org.openhab.binding.nest.internal.data.TopLevelStreamingData;
+import org.openhab.binding.nest.internal.listener.NestStreamingDataListener;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.gson.Gson;
+
+/**
+ * A client that generates events based on Nest streaming REST API Server-Sent Events (SSE).
+ *
+ * @author Wouter Born - Replace polling with REST streaming
+ */
+public class NestStreamingRestClient {
+
+    // Nest sends every 30 seconds a message to keep the connection alive
+    private static final long KEEP_ALIVE_MILLIS = Duration.ofSeconds(30).toMillis();
+
+    // Assume connection timeout when 2 keep alive message should have been received
+    private static final long CONNECTION_TIMEOUT_MILLIS = 2 * KEEP_ALIVE_MILLIS + KEEP_ALIVE_MILLIS / 2;
+
+    private static final String AUTH_REVOKED = "auth_revoked";
+    private static final String ERROR = "error";
+    private static final String KEEP_ALIVE = "keep-alive";
+    private static final String OPEN = "open";
+    private static final String PUT = "put";
+
+    private Logger logger = LoggerFactory.getLogger(NestStreamingRestClient.class);
+
+    private final List<NestStreamingDataListener> listeners = new CopyOnWriteArrayList<>();
+    private final EventSource eventSource;
+    private final Gson gson = new GsonBuilder().setDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'").create();
```

Comment:
Why do you need this ;-)

Reply:
Looking at the line numbers of your [stacktrace](https://github.com/openhab/openhab2-addons/pull/2737#issuecomment-333316022) I think you were testing with openHAB 2.1.0?

openHAB 2.1.0 uses Gson 2.3.1 (2014-11-20) which only supports a very simple ISO 8601 format:

```
  private static DateFormat buildIso8601Format() {
    DateFormat iso8601Format = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'", Locale.US);
    iso8601Format.setTimeZone(TimeZone.getTimeZone("UTC"));
    return iso8601Format;
  }
```

So it will fail to parse the milliseconds part of the timestamps, e.g.:
```
 "last_connection": "2017-09-30T11:45:55.599Z",
```

A better  ISO 8601 date type adapter was added in [Gson 2.4](https://github.com/google/gson/blob/master/CHANGELOG.md#version-24). 

I was testing with openHAB 2.2.0-SNAPSHOT which is using Gson 2.7.0 (2016-06-14) and also supports [more ISO 8601 timestamp variants](https://github.com/google/gson/blob/gson-parent-2.7/gson/src/main/java/com/google/gson/internal/bind/util/ISO8601Utils.java#L116).

Reply:
Then something could be messed up with upgrading **my** local docker, I upgrade it very regular and it is definitely on 2.2.

Reply:
@kaikreuzer / @sjka  I was really curious what was going on here with different runtime behaviour regarding GSON parsing, so I took a closer look. If I download the current openHAB snapshot and inflate that, it shows:

```
   ...
   creating: runtime/system/com/google/
   creating: runtime/system/com/google/guava/
   creating: runtime/system/com/google/guava/guava/
   creating: runtime/system/com/google/guava/guava/15.0/
   creating: runtime/system/com/google/guava/guava/18.0/
   creating: runtime/system/com/google/code/
   creating: runtime/system/com/google/code/findbugs/
   creating: runtime/system/com/google/code/findbugs/annotations/
   creating: runtime/system/com/google/code/findbugs/annotations/2.0.1/
   creating: runtime/system/com/google/code/gson/
   creating: runtime/system/com/google/code/gson/gson/
   creating: runtime/system/com/google/code/gson/gson/2.3.1/
   creating: runtime/system/com/google/code/gson/gson/2.7/
   ...
```

So we package multiple versions of GSON and Guava and maybe others which can will lead to different runtime behaviour for different people, is this a mistake or is this just an expected side effect?

Both version come from https://github.com/eclipse/smarthome/blob/master/features/karaf/esh-tp/src/main/feature/feature.xml